### PR TITLE
Repository-S3: Encryption using AWS kms key

### LINKED ISSUE #18561: Allow Other methods of s3 server-side encryption
**Describe the feature**: 

Currently, by setting server_side_encryption to true, the only method that is allowed is aes256. It would be great if we could specify other s3 encryption methods, e.g. aws:kms.

For aws:kms, if the bucket has a default key, it's just a matter of including the correct header, so one possibility may be an additional parameter to specify the value of the `x-amz-server-side-encryption` header to get sent with requests. But the ability to specify an ams key to use would also be great.

### PR DESCRIPTION
**Pull request by suman29**
Close #18561

Currently elasticsearch provides server side encryption of s3 repository only using AES256 algorithm. 
This pull request in an enhancement on top of that. It provides a functionality to use aws:kms algorithm to encrypt s3 repositories if aws_kms_key is provided. 

If no aws_kms_key is provided, it would default to it's beahviour of encrypting s3 repository files using AES256.

A curl request to take snapshot would look like this:

curl -XPUT 'http://localhost:9200/_snapshot/s3_repository?verify=false&pretty' -d'
 {
   "type": "s3",
   "settings": {
     "bucket": "my-bucket",
     "base_path": "es/test-snapshot",
     "region": "us-east-1",
     "server_side_encryption": true,
     "aws_kms_key": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
   }
 }'

### CODE REVIEW

File: plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/MockAmazonS3.java [hunk 1]
```diff
@@ -61,6 +62,8 @@
     // length of the input data is 100 bytes
     private byte[] byteCounter = new byte[100];
 
+    private Region region;
+
```
Comment (sunkuet02) [inline plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/MockAmazonS3.java]: Isn't the blank line is unnecessary ?

**Review by sunkuet02 — CHANGES_REQUESTED** [review]

File: plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/MockAmazonS3.java (hunk 1 above)
Comment (suman29) [inline plugins/repository-s3/src/test/java/org/elasticsearch/repositories/s3/MockAmazonS3.java]: Sure, I will remove it.

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 2]
```diff
@@ -37,6 +35,8 @@
 import org.elasticsearch.repositories.RepositoryException;
 import org.elasticsearch.repositories.blobstore.BlobStoreRepository;
 
+import java.io.IOException;
```
Comment (yaauie) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: again, picking around unrelated changes with `git add --patch` is hugely helpful in reducing noise.

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 3]
```diff
@@ -53,10 +53,14 @@
 
     static final String TYPE = "s3";
 
-    /** The access key to authenticate with s3. This setting is insecure because cluster settings are stored in cluster state */
+    /**
+     * The access key to authenticate with s3. This setting is insecure because cluster settings are stored in cluster state
```
Comment (yaauie) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: don't include unrelated changes

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 4]
```diff
@@ -53,10 +53,14 @@
...
+     */
     static final Setting<SecureString> ACCESS_KEY_SETTING = SecureSetting.insecureString("access_key");
 
-    /** The secret key to authenticate with s3. This setting is insecure because cluster settings are stored in cluster state */
+    /**
+     * The secret key to authenticate with s3. This setting is insecure because cluster settings are stored in cluster state
```
Comment (yaauie) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: don't include unrelated changes

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 5]
```diff
@@ -75,25 +79,31 @@
...
+
+    /**
      * Minimum threshold below which the chunk is uploaded using a single request. Beyond this threshold,
      * the S3 repository will use the AWS Multipart Upload API to split the chunk into several parts, each of buffer_size length, and
      * to upload each part in its own request. Note that setting a buffer size lower than 5mb is not allowed since it will prevents the
      * use of the Multipart API and may result in upload errors. Defaults to the minimum between 100MB and 5% of the heap size.
      */
     static final Setting<ByteSizeValue> BUFFER_SIZE_SETTING = Setting.byteSizeSetting("buffer_size", DEFAULT_BUFFER_SIZE,
-            new ByteSizeValue(5, ByteSizeUnit.MB), new ByteSizeValue(5, ByteSizeUnit.TB));
+        new ByteSizeValue(5, ByteSizeUnit.MB), new ByteSizeValue(5, ByteSizeUnit.TB));
```
Comment (yaauie) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: unrelated changes

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 6]
```diff
@@ -75,25 +79,31 @@
...
 
     /**
      * Big files can be broken down into chunks during snapshotting if needed. Defaults to 1g.
      */
     static final Setting<ByteSizeValue> CHUNK_SIZE_SETTING = Setting.byteSizeSetting("chunk_size", new ByteSizeValue(1, ByteSizeUnit.GB),
-            new ByteSizeValue(5, ByteSizeUnit.MB), new ByteSizeValue(5, ByteSizeUnit.TB));
+        new ByteSizeValue(5, ByteSizeUnit.MB), new ByteSizeValue(5, ByteSizeUnit.TB));
```
Comment (yaauie) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: unrelated changes

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 7]
```diff
@@ -130,7 +140,7 @@
      * Constructs an s3 backed repository
      */
     S3Repository(RepositoryMetaData metadata, Settings settings,
-                        NamedXContentRegistry namedXContentRegistry, AwsS3Service s3Service) throws IOException {
```
Comment (yaauie) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: unrelated changes

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 8]
```diff
@@ -153,12 +168,13 @@
         String storageClass = STORAGE_CLASS_SETTING.get(metadata.settings());
         String cannedACL = CANNED_ACL_SETTING.get(metadata.settings());
 
-        logger.debug("using bucket [{}], chunk_size [{}], server_side_encryption [{}], " +
-            "buffer_size [{}], cannedACL [{}], storageClass [{}]",
-            bucket, chunkSize, serverSideEncryption, bufferSize, cannedACL, storageClass);
+        logger.debug("using bucket [{}], chunk_size [{}], server_side_encryption [{}]," +
+                "server_side_encryption_key [{}], buffer_size [{}], cannedACL [{}], storageClass [{}]",
```
Comment (yaauie) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: even though this is just `debug`, logging an encryption key is worrisome

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java [hunk 9]
```diff
@@ -139,6 +149,11 @@
         }
 
         boolean serverSideEncryption = SERVER_SIDE_ENCRYPTION_SETTING.get(metadata.settings());
+        String serverSideEncryptionKey = AWS_KMS_KEY.get(metadata.settings());
+        if (serverSideEncryptionKey == null) {
```
Comment (yaauie) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Repository.java]: if the key isn't set, why set it to an empty string?

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java [hunk 10]
```diff
@@ -71,7 +76,7 @@
         SocketAccess.doPrivilegedVoid(() -> {
             if (client.doesBucketExist(bucket) == false) {
                 throw new IllegalArgumentException("The bucket [" + bucket + "] does not exist. Please create it before " +
-                                                   " creating an s3 snapshot repository backed by it.");
```
Comment (yaauie) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java]: unrelated changes

File: plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3OutputStream.java [hunk 11]
```diff
@@ -40,18 +40,21 @@
...
 
     private byte[] buffer;
     private int count;
     private long length;
 
     private int flushCount = 0;
 
-    S3OutputStream(S3BlobStore blobStore, String bucketName, String blobName, int bufferSizeInBytes, boolean serverSideEncryption) {
+    S3OutputStream(S3BlobStore blobStore, String bucketName, String blobName, int
+        bufferSizeInBytes, boolean serverSideEncryption,  String serverSideEncryptionKey) {
```
Comment (yaauie) [inline plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3OutputStream.java]: unconventional newline between type and parameter name

**Review by yaauie — COMMENTED** [review]
Hello.

First, I'm a community member and not part of Elastic; my opinions might not line up perfectly with Elastic's.

First, the PR contains a lot of unrelated changes, which add noise and make it hard to decipher which changes are intentional and which ones are side-effects of your editor's settings. By using `git add --patch` and selecting only the relevant hunks, and by previewing your diff before committing with `git diff --cached`,  you can reduce the noise and make it easier for your reviewers to have rational conversations about the effect of your changes.

Second, I think as it stands, this PR doesn't really have a case for any one object "owning" the encryption key; because the `S3BlobStore` is passed around to just about everything else that needs the key, it _may_ make sense to make it the owner, in which case a lot of the changes to constructors aren't needed since they already have a reference to the `S3BlobStore`. 

Third, only after careful review did it become clear that the "keys" that are being passed around are key-_id_s. Naming them in such a way that indicates that they are KMS key _id_s would be hugely helpful in understanding the risk involved with, say, having public accessors, logging them, etc.

**Comment by javanna** [comment]
@tlrx could you have a look at this please?

**Review by sunkuet02 — APPROVED** [review]

**Comment by tlrx** [comment]
@suman29 Sorry to come so late. Are you still interested to work on this?

If so, I think that @yaauie 's comments are worth to have a look. I also think that the aws_kms_key must be a secure settings like `access_key` and `secret_key` already are, ie a setting stored in the `elasticsearch-keystore` but read back using `SecureSetting.secureString()`.

**Comment by tlrx** [comment]
Thanks @suman29 for this contribution, we greatly appreciated it and I strongly apologize for the time it took me to look at it. 

We are still interested in such a feature but we don't have immediate plan to work on it. Also, adding the feature might require more work as it should be integrated with the Elasticsearch keystore. Since this pull request is a bit stalled I'm going to close it (again, sorry!). If anyone is interested in revived it please feel free to comment or +1.

**Comment by robsweet** [comment]
I'll +1 this.  Also, as the key is really just the ARN for the KMS key, and not the key itself, it doesn't seem like it needs to be a secure setting.

**Comment by danlsgiga** [comment]
I'm also +1 here. Agree with @robsweet, the key ARN is not a secret in any form, so having it out of the secure keystore is just fine.

**Comment by wknapik** [comment]
+1

**Comment by SkySonR** [comment]
Please, merge this

**Comment by wknapik** [comment]
@SkySonR https://github.com/elastic/elasticsearch/pull/41738

**Pull request closed without merging**